### DISCUSSION TITLE
Persistent File System API within AudioWorkletProcessor

### DISCUSSION THREAD

**Discussion by terryzfeng** [emscripten-core/emscripten#21085] [Q&A]
I'm working on a project that utilizes Emscripten and a Web Audio AudioWorklet that makes different sounds when reading files in a file system. I want the AudioWorklet Processor to be able to read files (currently using libc). Because `AudioWorkletProcessor` Emscripten module runs on a separate audio rendering thread, I can't update my file system directly from the main thread, I can only do it on the audio thread via `postMessage` which is bad for performance. I also need to persist my file system so I'm trying to use IDBFS + IndexedDB. 

How can I access the EM_ASM `FS` object (in C which is compiled in the AudioWorkletProcessor) on the Main Thread so I can see and make changes to the file system?

I've tried mounting an IDBFS folder so that my main thread could also read/write to IndexedDB in hopes that this would be possible with the following code:

```c
 void mountFileSystem() {
    EM_ASM(
      console.log("mountFileSystem");
      FS.mkdir("/working");
      FS.mount(IDBFS, {}, "/working");
      FS.createDataFile("/working", "test.txt", "Hello World!", true, true);
      FS.syncfs(true, function(err) {
        assert(!err);
      });
    );
  }
```

However, when doing this in C in the AudioWorkletProcessor, I get this error:

_Aborted(Assertion failed: IDBFS used, but indexedDB not supported)_

Is there a better approach for doing this?

**Comment by sbc100** [comment]
Is there some reason why you need to use AudioWorklet for your use case?  If you are simply playing files from the filesystem I think you can just use the normal emscripten audio capabilities, no?

**Reply by terryzfeng** [reply]
I'm working on a Web IDE for the [ChucK](https://chuck.stanford.edu) audio programming language. The language being written in C++ is compiled targeting the Web Audio API for audio processing done via the AudioWorklet Processor/Node. This custom audio processing node is [WebChucK](https://github.com/ccrma/webchuck). 

As I'm building an IDE for WebChucK to handle working uploading files in the browser, everything currently gets passed into the MEMFS file system. It's easy to get files in, but harder to get files out; WebChucK can also write files of it's own to the file system. Ideally, I'm hoping for my Emscripten compiled AudioWorklet to be able to mount some kind of file system so that I can do FileIO on the main thread, view internal files, persist and capture them to save projects that users work on. 

WebChucK IDE can be found here: https://chuck.stanford.edu/ide

**Comment by sbc100** [comment]
@juj, you might have more advice on the AudioWorklet front.
